python_modules/libraries/dagster-gcp/dagster_gcp/bigquery/io_manager.py

```diff
@@ -0,0 +1,226 @@
+        service account key as the "gcp_credentials" configuration. Dagster will store this key in a
+        temporary file and set GOOGLE_APPLICATION_CREDENTIALS to point to the file. After the run completes,
+        the file will be deleted, and GOOGLE_APPLICATION_CREDENTIALS will be unset.
+
+    """
+
+    @io_manager(
+        config_schema={
+            "dataset": Field(
+                StringSource,
+                description=(
+                    "Name of the BigQuery dataset to use. If not provided, the last prefix before"
+                    " the asset name will be used."
+                ),
+                is_required=False,
+            ),
+            "project": Field(StringSource, description="The GCP project to use."),
+            "location": Field(
+                Noneable(StringSource),
+                is_required=False,
+                default_value=None,
+                description="The GCP location.",
+            ),
+            "gcp_credentials": Field(
+                Noneable(StringSource),
+                is_required=False,
+                default_value=None,
+                description=(
+                    "GCP authentication credentials. If provided, a temporary file will be created"
+                    " with the credentials and GOOGLE_APPLICATION_CREDENTIALS will be set to the"
+                    " temporary file."
+                ),
+            ),
+        }
+    )
+    def bigquery_io_manager(init_context):
+        """I/O Manager for storing outputs in a BigQuery database
+
+        Assets will be stored in the dataset and table name specified by their AssetKey.
+        Subsequent materializations of an asset will overwrite previous materializations of that asset.
+        Op outputs will be stored in the dataset specified by output metadata (defaults to public) in a
+        table of the name of the output.
+
+        Note that the BigQuery config is mapped to the DB IO manager table hierarchy as follows:
+        * project -> database
+        * dataset -> schema
+        * table -> table
+        """
+        temp_file_name = ""
+        if init_context.resource_config.get("gcp_credentials"):
+            with tempfile.NamedTemporaryFile("w", delete=False) as f:
+                f.write(init_context.resource_config.get("gcp_credentials"))
+                temp_file_name = f.name
+                os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = temp_file_name
+
+        yield DbIOManager(
+            type_handlers=type_handlers,
+            db_client=BigQueryClient(),
+            io_manager_name="BigQueryIOManager",
+            database=init_context.resource_config["project"],
+            schema=init_context.resource_config.get("dataset"),
+        )
+        if init_context.resource_config.get("gcp_credentials"):
+            os.environ.pop("GOOGLE_APPLICATION_CREDENTIALS", None)
+            os.remove(temp_file_name)
```

Comment:
Why use `delete=False` and remove it here instead of putting the `yield` inside the `with` statement?

Reply:
In most cases the user should be authing with gcp via the CLI `gcloud auth application-default login` or by setting `GOOGLE_APPLICATION_CREDENTIALS` to a file containing the key. But in cases like serverless, there's no way for the user to upload the key file to the container, so we need to let them provide creds via config. The `yield` is outside with `with` because the `with` only happens in the case when we're authing with credentials in the config 

I suppose we could do this instead:

```
        mgr = DbIOManager(
            type_handlers=type_handlers,
            db_client=BigQueryClient(),
            io_manager_name="BigQueryIOManager",
            database=init_context.resource_config["project"],
            schema=init_context.resource_config.get("dataset"),
        )
        temp_file_name = ""
        if init_context.resource_config.get("gcp_credentials"):
            with tempfile.NamedTemporaryFile("w") as f:
                f.write(init_context.resource_config.get("gcp_credentials"))
                temp_file_name = f.name
                os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = temp_file_name

                try:
                    yield mgr
                finally:
                    os.environ.pop("GOOGLE_APPLICATION_CREDENTIALS", None)

        return mgr

Reply:
tried this and having the yield of the manager inside the with statement means that the contents of the key isn't fully written to the file. I don't quite understand why yet, but basically it seems like the key is only "saved" to the file once the file is closed. so i think we need to keep the `delete=False` and manually delete later